In [11]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn import metrics
from sklearn.metrics import make_scorer, mean_squared_error

In [12]:
# read in the advertising dataset
data = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Advertising.csv', index_col=0)

In [13]:
data.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [14]:
feature_cols = ["TV", "radio","newspaper"]
X = data[feature_cols]
y = data['sales']

We will use mean_squared as the cross validation function because we can't use accuracy since lm is not a classification model.
We want to use root MSE but it's not available directly so we will use MSE and root it.

Also, mean_squared error is not a scoring function now, to get that we have to use make_scorer and then use the mse inside it. more about make_scorer - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html

In [17]:
lm = LinearRegression()
scorer = make_scorer(mean_squared_error, greater_is_better=True)
scores = cross_val_score(lm,X,y,cv=10, scoring = scorer)
scores

array([3.56038438, 3.29767522, 2.08943356, 2.82474283, 1.3027754 ,
       1.74163618, 8.17338214, 2.11409746, 3.04273109, 2.45281793])

In [21]:
rmse_scores = np.sqrt(scores)
rmse_scores

array([1.88689808, 1.81595022, 1.44548731, 1.68069713, 1.14139187,
       1.31971064, 2.85891276, 1.45399362, 1.7443426 , 1.56614748])

In [23]:
rmse_scores.mean()

1.6913531708051792

So the RMSE of the lm model using 3 factors gave 1.69. NOw let's check it using only 2, ie, removing newspaper

In [25]:
feature_cols1 = ["TV", "radio"]
X = data[feature_cols1]
y = data['sales']

In [26]:
lm = LinearRegression()
scorer = make_scorer(mean_squared_error, greater_is_better=True)
scores = cross_val_score(lm,X,y,cv=10, scoring = scorer)
scores

array([3.38264572, 3.28736055, 2.09185507, 2.81784389, 1.27311083,
       1.74391853, 8.0006111 , 2.07650936, 3.03635301, 2.4289835 ])

In [28]:
rmse_scores_mean = np.sqrt(cross_val_score(lm,X,y,cv=10, scoring = make_scorer(mean_squared_error))).mean()
rmse_scores_mean

1.6796748419090766

We can use the single line code by using np.sqrt, .mean(), and make_scorer in the same line
Now we can see that the rmse has reduced to 1.68, which is better than 1.69. So we should use only 2 factors in this case

# Improvements in CrossValidation
        Repeated Cross Validation
        Creating a hold out set
        Doing all feature engineering within CV